### PR TITLE
Remove --port and update --publish for services to support syntaxes

### PR DESCRIPTION
**Pull request by vdemeester**
This is part of #28527 — "The newly introduced `--port` flag should likely disappear in favor of `--publish`, ensuring that the short form maps cleanly to the long form"

* Removing `--port`, `--port-add`, `--port-rm` flags
* Add support for simple and complex syntax on `--publish` flag (and `--publish-add`, `--publish-rm`), using `PortOpt`
* Validate the changes on `PortOpt` with unit tests.

```
$ docker service create --name t1 --publish 81:81/tcp --publish target=80,published=80,protocol=tcp,mode=ingress busybox top
# […]
$ docker service update --publish-add 82 --publish-rm 80:80/tcp --publish-add target=83 --publish-rm target=80 t1
# […]
```

- [ ] Probably need a small refactoring to share the short/long syntax detection
- [x] Add one integration test to validate it works (but not all possible cases, covered by unit tests)

/cc @thaJeztah @icecrime @dnephin @mavenugo 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by icecrime** [comment]
Cc @vieux: I'm not sure if you want to take this for 1.13.0.

File: cli/command/service/opts.go [hunk 1]
```diff
@@ -649,9 +619,6 @@ const (
 	flagPublish               = "publish"
 	flagPublishRemove         = "publish-rm"
 	flagPublishAdd            = "publish-add"
-	flagPort                  = "port"
-	flagPortAdd               = "port-add"
-	flagPortRemove            = "port-rm"
```
Comment (dnephin) [inline cli/command/service/opts.go]: I don't think we can just remove `--port` they were in the 1.12 release as non-experimental. We have to deprecate them.

Comment (vdemeester) [inline cli/command/service/opts.go]: @thaJeztah you are too quick for me :sweat_smile:

**Comment by LK4D4** [comment]
LGTM
ping @vieux

File: opts/port_test.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+				{
+					Protocol:      "tcp",
+					TargetPort:    80,
+					PublishedPort: 8080,
+					PublishMode:   "host",
+				},
+			},
+		},
+		{
+			value: "target=80,published=8080,mode=host",
```
Comment (mavenugo) [inline opts/port_test.go]: Good tests. Can you also pls include a "ingress" mode test ?

File: opts/port_test.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+func TestPortOptValidComplexSyntax(t *testing.T) {
+	testCases := []struct {
+		value    string
+		expected []swarm.PortConfig
+	}{
+		{
+			value: "target=80",
+			expected: []swarm.PortConfig{
+				{
+					TargetPort: 80,
```
Comment (mavenugo) [inline opts/port_test.go]: Is there a reason to not check for other default states such as `Protocol`, `PublishMode` ?
I think it is important to make sure the default protocol (tcp) and publishmode (ingress) are parsed appropriately

**Review by mavenugo — COMMENTED** [review]
Thanks @vdemeester 
minor comments on tests. I will try it out and give more comments if any.

File: cli/command/service/opts.go [hunk 4]
```diff
@@ -459,16 +431,14 @@ func newServiceOptions() *serviceOptions {
...
-			publishPorts: opts.NewListOpts(ValidatePort),
-		},
-		groups:    opts.NewListOpts(nil),
-		logDriver: newLogDriverOptions(),
-		dns:       opts.NewListOpts(opts.ValidateIPAddress),
-		dnsOption: opts.NewListOpts(nil),
-		dnsSearch: opts.NewListOpts(opts.ValidateDNSSearch),
-		hosts:     opts.NewListOpts(runconfigopts.ValidateExtraHost),
-		networks:  opts.NewListOpts(nil),
+		endpoint:        endpointOptions{},
```
Comment (dnephin) [inline cli/command/service/opts.go]: This line is unnecessary now

**Review by dnephin — APPROVED** [review]
LGTM

File: cli/command/service/opts.go (hunk 4 above)
Comment (vdemeester) [inline cli/command/service/opts.go]: Oh right :sweat_smile:

File: opts/port_test.go (hunk 3 above)
Comment (vdemeester) [inline opts/port_test.go]: I should have added comment. This one is just about the required field (`target` here). The default protocol and publish mode are handled server side so the client doesn't send them (that's why the `Swarm.PortConfig` is that plain).

File: opts/port_test.go (hunk 2 above)
Comment (vdemeester) [inline opts/port_test.go]: ah you're rihgt, I thought there was one, my bad :sweat_smile:

Comment (vdemeester) [inline opts/port_test.go]: added :angel:

**Comment by mavenugo** [comment]
Thanks @vdemeester 

LGTM

**Review by mavenugo — APPROVED** [review]

**Comment by thaJeztah** [comment]
@vdemeester can you also update the completion scripts?

https://github.com/docker/docker/blob/9ec471049d4a435a489fa27767faebadb0f66e3c/contrib/completion/bash/docker#L2755

https://github.com/docker/docker/blob/d23bb3d0831fa200d45510e9f85b0086b2533553/contrib/completion/zsh/_docker#L1798

Should we update the changelog as part of this PR as well? (I suggest to add both the "old" and "new" PR as a link); https://github.com/docker/docker/blob/caaa52c124539acb396c0b711ab95bc7ed459540/CHANGELOG.md#L59

**Comment by vdemeester** [comment]
ah right :angel:

**Comment by vdemeester** [comment]
@thaJeztah updated :angel:

**Comment by mavenugo** [comment]
@vdemeester rebase needed :(

**Comment by thaJeztah** [comment]
Found one issue; if defaults are omitted during `service update`, the previous value is not changed;

publish on port 8080, "ingress" mode

```
docker service create --publish published=8080,target=80,protocol=tcp --name web nginx:alpine

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"tcp","TargetPort":80,"PublishedPort":8080,"PublishMode":"ingress"}]
```

change to "host" mode

```
docker service update --publish-rm 80 --publish-add published=8080,target=80,protocol=tcp,mode=host web

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"tcp","TargetPort":80,"PublishedPort":8080,"PublishMode":"host"}]
```

change back to "ingress" mode

```
docker service update --publish-rm 80 --publish-add published=8080,target=80,protocol=tcp web

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"tcp","TargetPort":80,"PublishedPort":8080,"PublishMode":"host"}]
```




publish on port 8080, "tcp" (default)

```
docker service create --publish published=8080,target=80 --name web nginx:alpine

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"tcp","TargetPort":80,"PublishedPort":8080,"PublishMode":"ingress"}]
```

change to "udp"

```
docker service update --publish-rm 80 --publish-add published=8080,target=80,protocol=udp web

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"udp","TargetPort":80,"PublishedPort":8080,"PublishMode":"ingress"}]
```

change back to "tcp" (default)

```
docker service update --publish-rm 80 --publish-add published=8080,target=80 web

docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"udp","TargetPort":80,"PublishedPort":8080,"PublishMode":"ingress"}]
```

**Comment by vdemeester** [comment]
So… this is conflicting with https://github.com/docker/docker/pull/25860 :sweat_smile: I temporarly disabled the validation and added comment because I'm not sure about what to do…

**Comment by vdemeester** [comment]
Please look closely at https://github.com/docker/docker/pull/28943/commits/6abada0ee70ea6b09a12c265ce1d1db832861890 (this fixes @thaJeztah's comment **BUT** changes a little bit the UX to something that feels more natural).

With that commit, `--publish-add 8081:81 --publish-add 8082:82 --publish-rm 80 --publish-rm 81/tcp --publish-rm 82/ucp` would thus result in 81 and 82 to be published (instead of only 82 currently).

/cc @dnephin @thaJeztah @icecrime @mavenugo

**Comment by vdemeester** [comment]
Rebased 👼

**Comment by thaJeztah** [comment]
Testing my previous examples, I ran into this; adding a port in mode "host" cannot be removed using the "simple" syntax

```
$ docker service create --publish published=8080,target=80,protocol=tcp,mode=host --name web nginx:alpine
oq5jnkhug6hjddtewift80dz9

$ docker service update --publish-rm 80 web
web

$ docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
[{"Protocol":"tcp","TargetPort":80,"PublishedPort":8080,"PublishMode":"host"}]
```

This worked;

```
$ docker service update --publish-rm target=80,mode=host web
web

$ docker service inspect --format '{{ json .Spec.EndpointSpec.Ports }}' web
null
```

**Comment by mavenugo** [comment]
@thaJeztah @vdemeester @vieux we need this for 1.13.0 and I think @thaJeztah's UX observation is expected. Do we have any concerns with this PR still to be merged and cherry-picked ?

**Comment by thaJeztah** [comment]
@mavenugo I'm good with this PR, just wanted confirmation that we agree on the change in behavior, i.e.;

- previously, when having both `--publish-rm`, and `--publish-add`, `--publish-rm` would always be executed _last_. This meant that you could not "replace" a port in a single update.
- the new change always performs `--publish-rm` _first_, then _adds_ the new ports specified in `--publish-add`.

I think this was an oversight in the original implementation, and I cannot think of a practical use of _adding_ a new port and _removing_ it in the same update. Similarly, I cannot think of a case where this would be a real "breaking" change.

If we agree on this, I suggest we look at other `-rm` and `-add` flags, which currently may use the "wrong" behavior as well (at least `secrets`)

**Comment by mavenugo** [comment]
@thaJeztah I agree. the current behaviour is more correct than the previous one.
This PR already has my LGTM.

**Comment by thaJeztah** [comment]
ok, let's go for it 👍

**Pull request merged**